python/graphstorm/gconstruct/utils.py

```diff
@@ -979,3 +999,77 @@ def partition_graph(g, node_data, edge_data, graph_name, num_partitions, output_
         # the new_edge_mapping contains per edge type on the ith row
         # the original edge id for the ith edge.
         save_maps(output_dir, "edge_mapping", new_edge_mapping)
+
+def get_hard_edge_negs_feats(hard_edge_neg_ops):
+    """ Get feature names of hard negatives for each edge type.
+
+        Parameters
+        ----------
+        hard_edge_neg_ops: HardEdgeNegativeTransform
+            A list of edge hard negative transformations.
+    """
+    hard_edge_neg_feats = {}
+    for hard_edge_neg_op in hard_edge_neg_ops:
+        edge_type = hard_edge_neg_op.target_etype
+        neg_ntype = hard_edge_neg_op.neg_ntype
+
+        if edge_type not in hard_edge_neg_feats:
+            hard_edge_neg_feats[edge_type] = {neg_ntype: [hard_edge_neg_op.feat_name]}
+        else:
+            if neg_ntype in hard_edge_neg_feats[edge_type]:
+                hard_edge_neg_feats[edge_type][neg_ntype].append(hard_edge_neg_op.feat_name)
+            else:
+                hard_edge_neg_feats[edge_type][neg_ntype] = [hard_edge_neg_op.feat_name]
+
+    return hard_edge_neg_feats
+
+def shuffle_hard_nids(data_path, num_parts, hard_edge_neg_feats):
```

Comment:
do you mean shuffle or map?

Reply:
Shuffle the nids according to graph partition.